### ISSUE TITLE
[java] Make CompareObjectWithEquals allow comparing against constants

### ISSUE THREAD

**Issue by oowekyala** [pmd/pmd#3205]
**Affects PMD Version:** 6.33

**Rule:**  [CompareObjectsWithEquals](https://pmd.github.io/latest/pmd_rules_java_errorprone.html#compareobjectswithequals)

**Description:** CompareObjectsWithEquals would be less noisy if it would accept comparing against final fields which serve as sentinel values

**Code Sample demonstrating the issue:**

```java
class MyClass {
    static final MyClass MISSING = new MyClass();

    public static void isMissing(MyClass obj) {
          return obj == MISSING; // no violation expected...
    }
}
```

Also see
* 7a058e12df6204eb4717fb6d8b58cf1a559d0379
* In PMD 7, JTypeMirror is designed to use several sentinel values which are final fields of [TypeSystem](https://github.com/pmd/pmd/blob/c35b4e879ec1ac22b392e41c6e650da8fa46fbdf/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/types/TypeSystem.java#L61-L62), so we have to suppress the rule a lot: eg [TypeOps](https://github.com/pmd/pmd/blob/c35b4e879ec1ac22b392e41c6e650da8fa46fbdf/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/types/TypeOps.java#L49-L50), [ExprOps](https://github.com/pmd/pmd/blob/pmd/7.0.x/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/types/internal/infer/ExprOps.java#L38-L39), TypeSystem above...
* https://github.com/pmd/pmd/blob/c9a463bd4fb723ec43e284ba282e34ebb64b4967/pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/internal/StreamImpl.java#L122
* https://github.com/pmd/pmd/blob/d079e0841e5966b0f8bdb23c599f5ccb70330bc7/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/types/internal/infer/Infer.java#L176

Note that in many of those cases the constant is not static. So we should allow non-static constants, possibly, provided they have an all uppercase name.

**Expected outcome:**

*   Does PMD report a violation, where there shouldn't be one? -> false-positive

**Running PMD through:** Any
